Add actuator endpoint for finding and deleting sessions

vpavic opened the pull request:
This PR is an alternative approach to implementing sessions endpoint to one proposed in #5730.

Key differences:

- unlike #5730, this implementation doesn't provide `Endpoint` implementation i.e. it only provides MVC and JMX endpoints that simply delegate to Spring Session's `FindByIndexNameSessionRepository` (this is analogous to audit events endpoints - see #6579)
- JMX endpoint is based on support for `JmxEndpoint` implementations without a backing `Endpoint` (introduced in #6579)
- Actuator docs

vpavic commented:
I've adapted this PR to the new Actuator endpoint infrastructure. Hands-on developer experience with the new stuff is just great!

snicoll commented on spring-boot-actuator/src/main/java/org/springframework/boot/actuate/session/SessionsWebEndpointExtension.java:
```diff
@@ -0,0 +1,49 @@
+	@ReadOperation
+	public WebEndpointResponse<SessionsReport> sessionsForUsername(String username) {
+		if (username == null) {
+			return new WebEndpointResponse<>(HttpStatus.BAD_REQUEST.value());
```
You can't use `HttpStatus` here as this class may not be available when running with Jetty. That's why that object is taking the raw `int` code.

vpavic replied:
Web framework neutrality completely slipped my mind when using `HttpStatus`, I'll address that shortly.

Note that I used the same pattern in `AuditEventsWebEndpointExtension` that was introduced in #10322. Quick search also revealed another usage in `HeapDumpWebEndpoint`.

snicoll replied:
Thanks for the feedback. I've created #10348

snicoll commented on spring-boot-actuator/src/test/java/org/springframework/boot/actuate/session/SessionsEndpointTests.java:
```diff
@@ -0,0 +1,88 @@
+
+	@Test
+	public void deleteSession() {
+		this.endpoint.deleteSession(session.getId());
```
That test does nothing. Verify that the repository was called with the right argument would be in order:

```
verify(this.repository).deleteById(session.getId());
```

snicoll commented on spring-boot-docs/src/main/asciidoc/production-ready-features.adoc:
```diff
@@ -106,6 +106,10 @@ The following technology agnostic endpoints are available:
 
+|`sessions`
+|Allows retrieving user's sessions from Spring Session backed session store, and deleting
+them.
```
I'd add that to the first part of the sentence (see above for an example).

vpavic commented:
Thanks for the review @snicoll, I've addressed your concerns.

Regarding the sample remark, would it be OK to reuse the existing `spring-boot-sample-session-redis` to showcase this functionality?

snicoll commented:
Yes. Perhaps we could first rename it and use something else. jdbc comes to mind even I know you guys don't want to promote it too much?

That sample looks outdated as it requires a redis server.

vpavic commented:
Can we use the similar approach to one in `spring-boot-sample-cache` and have multiple build profiles?

snicoll commented:
Maybe. Not sure that's an immediate goal but I am happy to review a PR

vpavic commented:
@snicoll I've opened #10351 to improve the sample.

vpavic commented:
I've updated the PR to adopt to changes from #10350 and enable the new endpoint in `spring-boot-sample-session`.

vpavic commented:
Transient errors involving `WebFluxEndpointIntegrationTests` like the one below are quite often on Travis CI lately. It's been basically 50-50 with my PR updates over the past few days.

```
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   WebFluxEndpointIntegrationTests>AbstractWebEndpointIntegrationTests.writeOperation:165->AbstractWebEndpointIntegrationTests.load:312->AbstractWebEndpointIntegrationTests.load:300->AbstractWebEndpointIntegrationTests.lambda$load$24:313->AbstractWebEndpointIntegrationTests.lambda$writeOperation$11:170 » IllegalState
[INFO] 
[ERROR] Tests run: 440, Failures: 0, Errors: 1, Skipped: 1
```

wilkinsona commented:
We see those on Bamboo too. @bclozel has been investigating. My suspicion is that it's a Linux-specific problem in Netty or Reactor Netty.

vpavic commented:
Thanks for the feedback. FWIW I didn't hit those problems on Ubuntu 17.04.

manderson23 commented:
@vpavic was there a reason that the endpoint doesn't have the option of returning the IDs of all sessions within the repository?

wilkinsona commented:
There's no support in Spring Session for getting the IDs of all sessions known to the repository.
